backend/tests/test_llm_integration.py

```
@@ -78,5 +78,49 @@ def test_get_embeddings_routing(self, mock_get_emb):
+        mock_client = MagicMock()
+        mock_client.invoke.side_effect = [Exception("transient"), "result"]
+        result = _invoke_with_retry(mock_client, ["msg"])
+        self.assertEqual(result, "result")
+        self.assertEqual(mock_client.invoke.call_count, 2)
+        mock_sleep.assert_called_once_with(1)  # 2**0 = 1
+
+    @patch('time.sleep')
+    def test_raises_after_all_retries_exhausted(self, mock_sleep):
+        mock_client = MagicMock()
+        mock_client.invoke.side_effect = RuntimeError("permanent error")
+        with self.assertRaises(RuntimeError):
```

Ruff (0.15.20)

[warning] Use `pytest.raises` instead of unittest-style `assertRaises`

Replace `assertRaises` with `pytest.raises`

(PT027)